[CBRD-26243] (Phase-1) Scale Range Expansion (1/2)

jongmin-won opened the pull request:
http://jira.cubrid.org/browse/CBRD-26243

### Purpose
#### 1. NUMERIC 최대 범위 (=입력 리터럴 허용 범위)
```
1) 지수 표현
1.0 × 10^-252 <= |Value| < 1.0 * 10^254

2) Precision/ Scale
1 <= prec <= 43
-211 <= scale <= 252
```
- 0 허용 (즉, |Value| = 0 가능)
- 리터럴 길이 제한
  - 부호(+/−)는 길이에 포함되지 않고, "0."(소수점 앞의 0과 소수점)은 길이에 포함되며, 전체 길이는 254자 이내까지만 허용.
  - 254자를 초과하면 오류 없이 잘라내어 254자까지만 처리한다.
  - 예: -123.45는 숫자(5)+소수점(1)=6자
  - 예: +0.0012는 0(1) + 소수점(1) + 숫자(4)= 6자

- 리터럴 값의 NUMERIC 변환 규칙
  - 리터럴 값을 NUMERIC으로 변환할 때, 부호(+/−)와 "0."(소수점 앞의 0과 소수점)은 제외하고, 남은 숫자로 precision/scale을 계산함.
  - 예: -123.45는 prec : 5, scale 2
  - 예: +0.0012는 prec : 2, scale 4

- 부호가 제외된 이유
  - 단, String을 NUMERIC으로 캐스팅할 경우 부호는 포함 됩니다.
  - 예 : SELECT cast('+12000.123456' as numeric);
```
1) 양수 값  (= + 부호 포함)
- Node : PT_SELECT -> PT_VALUE
- PT_VALUE 노드에는 부호 없는 리터럴 값이 저장됩니다.

2) 음수 값
- Node : PT_SELECT -> PT_EXPR -> PT_VALUE 
- PT_EXPR의 op가 PT_UNARY_MINUS로 설정됩니다.
- PT_VALUE 단계에서는 부호 없는 리터럴 값을 처리하고, 이후 PT_EXPR에서 부호가 적용됩니다.
```

#### 2. Fixed Numeric
```
numeric(p,s)
```
- p (precision): 유효 자릿수.
  - 가장 왼쪽의 0이 아닌 숫자부터 가장 오른쪽 유효 숫자까지의 자릿수
  - 범위 : 1 ~ 38
  - 16바이트 base-256 정밀도로, 최대 38자리 십진수 표현 가능.
- s (scale): 소수점 기준 소수부 자릿수.
  - 범위 : -84 ~ 127
  - 음수 scale은 소수점 왼쪽에서 반올림 단위를 지정
- 전체 허용 범위 (부호 무관)
```
1) 지수 표현
1.0 * 10^−127 <= |Value| < 1.0 * 10^122

2) Precision/ Scale
1 <= prec <= 38
-84 <= scale <= 127
```

### Implementation
#### 1. Numeric Parsing 로직 수정
##### 1-1. 정수부와 소수부를 분리하여 파싱.
##### 1-2. Oracle 방식처럼 prec/scale 을 일관되게 해석
- prec  : 가장 왼쪽의 0이 아닌 숫자부터 가장 오른쪽 유효 숫자까지의 자릿수
- scale : 소수점 기준으로, 소수부의 시작부터 가장 오른쪽 유효 숫자까지의 자릿수
##### 1-3. 최대 유효 숫자 초과 시 반올림
- 입력된 literal의 길이가 유효 숫자(precision) 자릿수를 초과하는 경우, Numeric Parsing 단계에서 반올림을 수행합니다.
  - Phase-1 최대 유효 숫자: 38자리 (16바이트)
    - literal의 유효 숫자가 38 자리를 초과하는 경우, 39번째 자리에서 반올림하여 38자리까지 유지.
  - Phase-3 최대 유효 숫자: 43자리 (18바이트)
    -  literal의 유효 숫자가 43 자리를 초과하는 경우, 44번째 자리에서 반올림하여 43자리까지 유지.
- Numeric Parsing 단계에서 반올림을 수행하는 이유
  - Oracle 공식 매뉴얼에 명시되어 있지는 않지만, 유효 숫자 범위를 초과하는 literal이 입력되면 Oracle은 허용 가능한 범위까지만 반올림하여 값을 보여줍니다.
  - CUBRID도 동일한 동작을 보장하기 위해 Numeric Parsing 단계에서 반올림을 적용합니다.
```sql
select 42.909999999999999999999999999999999999999;

- Oracle: 42.91
- CUBRID (반올림 미적용): 42.909999999999999999999999999999999999
- CUBRID (반올림 적용): 42.910000000000000000000000000000000000
```
##### 5. trailing zero 제거 제외
- Phase-1 에서는 Numeric Paring에서 trailing zero 제거를 제외하도록 수정
- 단, oracle_compat_number_behavior 파라미터를 yes로 설정한 경우에는 기존과 동일하게 trailing zero를 제거하여 출력

#### 2. 매크로 확장
##### 2-1.  NUMERIC_MAX_STRING_SIZE : 81 -> 512
- NUMERIC 값을 위한 내부 작업용 버퍼의 최대 크기를 정의함.
- 이 버퍼는 NUMERIC 처리 전반에 걸쳐 광범위하게 사용됨 (예시 일부)
  - NUMERIC 결과의 출력 표현
  - 산술 연산 및 내장 함수 처리
  - 스케일 조정(소수 자릿수 정렬) 및 반올림
- 버퍼 크기는 TWICE_NUM_MAX_PREC(256) 값의 2배 로 설정되며, 이는 스케일 조정 과정에서 발생할 수 있는 자릿수 증가를 위함
##### 2-2. TWICE_NUM_MAX_PREC : 76 -> 256
- NUMERIC 값에 저장할 수 있는 최대 유효 숫자 자릿수(부호와 소수점 제외)를 정의함.
- NUMERIC 값의 최대 범위를 내부적으로 표현하기 위한 버퍼 크기로 사용됩니다.
- 또한, base-256 값을 10진 문자열로 변환할 때 사용됩니다.
  - 최대 자릿수 (43 + 211) = 254 + 2 (여분 자릿수) = 256
  - 항상 NUMERIC_MAX_STRING_SIZE보다 작아야 합니다.

### Remarks
#### 1. Numeric Parsing에서 DB_MAX_NUMERIC_PRECISION(43) 대신 임시 변수 phase_1_tmp_max_prec(38)를 사용합니다.
- 현재(Phase-1)는 NUMERIC의 최대 범위 확장만 적용된 상태이므로, DB_MAX_NUMERIC_PRECISION(43)을 그대로 사용할 경우 올바르게 동작하지 않습니다.
- 따라서 아직 NUMERIC의 전체 범위 (43,-211)은 지원되지 않으며, (1,252) <= (prec, scale) <= (38,-211) 범위까지만 정상 동작합니다.

Commit:
1. Scale 범위를 -84 ~ 127로 확장
2. Numeric Parsing 로직 수정 (Phase-3에서 Precision 최대 40 확장 예정, 단 테이블 생성 시 Precision은 기존과 동일하게 38까지 허용)
* 정수부와 소수부를 분리하여 파싱.
* Precision <= 38
** 기존 NUMERIC 방식 적용 (정수부+소수부 자릿수 합산 기준).
* Precision > 38
** Oracle 방식 적용 (왼쪽부터 0이 아닌 첫 유효 숫자부터 끝까지 카운트)
* Parsing 규칙
** 소수 포함 전체 자릿수가 38자리를 초과하는 경우 -> 39자리에서 반올림하여 38자리로 저장 및 출력
** 정수 부분이 38자리를 초과하는 경우 -> 반올림 없이 Overflow 에러 발생
* trailing zero 제거
** Parsing 과정에서 반올림한 결과 scale이 존재하는 경우, 소수부의 trailing zero를 제거

Commit:
code-style

ctshim commented on msg/ko_KR.utf8/cubrid.msg:
```diff
@@ -1964,6 +1967,7 @@ $set 8 MSGCAT_SET_PARSER_SEMANTIC
 330 저장 프로시저/함수 인자 디폴트 값의 문자열 표현은 255자를 넘을 수 없습니다.
 331 Static SQL에 사용된 호스트 표현식은 변수나 레코드 필드이어야 합니다: '%1$s'.
 
+332 NUMERIC의 스케일 범위(-84 에서 127)를 초과했습니다.
```
332 NUMERIC의 스케일 범위(-84부터 127까지)를 초과했습니다.

jongmin-won replied:
코드 리뷰 감사합니다.
말씀하신 부분을 반영하여 수정했습니다.

Commit:
Fix code review feedback

Commit:
Numeric Parsing 단계에서 trailing zero를 제거하지 않도록 수정한다.

ctshim commented on src/query/query_opfunc.c:
```diff
@@ -760,7 +760,7 @@ qdata_coerce_dbval_to_numeric (DB_VALUE * dbval_p, DB_VALUE * result_p)
   DB_DATA_STATUS data_stat;
 
-  db_value_domain_init (result_p, DB_TYPE_NUMERIC, DB_DEFAULT_PRECISION, DB_DEFAULT_SCALE);
+  db_value_domain_init (result_p, DB_TYPE_NUMERIC, DB_DEFAULT_PRECISION, DB_DEFAULT_NUMERIC_SCALE);
```
`DB_DEFAULT_SCALE`로 되어 있던 것은 기존 코드에서도 일종의 오류였던 건가요?
아니면 이번 수정과 관련되어 고칠 필요가 있었던 건가요?

jongmin-won replied:
기존 코드 자체의 오류는 아닙니다.

numeric 타입의 초기 domain 설정 과정에서 DB_DEFAULT_PRECISION(-1), DB_DEFAULT_SCALE(-1) 인 경우에는
기본적으로 numeric(15,0)으로 설정되도록 되어 있습니다.

DB_DEFAULT_SCALE(-1)을 DB_DEFAULT_NUMERIC_SCALE(0)으로 수정한 이유는 다음과 같습니다.
- 음수 scale이 허용되면서 사용자가 -1을 의도적으로 지정할 수 있음
- 따라서 db_value_domain_init() 내부에서 DB_DEFAULT_SCALE(-1)을 무조건 DB_DEFAULT_NUMERIC_SCALE(0)으로 바꾸는 로직을 수정할 필요가 있음

만약 db_value_domain_init()을 기존 그대로 둘 경우, numeric의 기본 타입이 (15,0)이 아니라 (15,-1)로 설정되어 의도하지 않은 결과가 발생하게 됩니다.

이를 방지하기 위해, db_value_domain_init()을 사용하는 함수에서 타입이 NUMERIC인 경우에는 DB_DEFAULT_SCALE(-1)을 DB_DEFAULT_NUMERIC_SCALE(0)으로 설정하도록 수정했습니다.

hyunikn replied:
위에 비슷한 질문을 드렸는데, 이 답변을 보니까 이해가 됩니다. 

그런데, DB_DEFAULT_SCALE 을 numeric scale 이 가질 수 없는 -85 이하의 수로 바꾸면 
위 변경 중에서 DB_DEFAULT_SCALE 대신 DB_DEFAULT_NUMERIC_SCALE 로 바꾼 부분들을 원복할 수 있을 것 같습니다.

jongmin-won replied:
리뷰 의견을 반영하여 DB_DEFAULT_SCALE 값을 -1에서 -9999로 변경하였으며,
이전에 제거했던 DB_DEFAULT_SCALE 관련 조건들도 모두 원복했습니다.

ctshim commented on src/query/numeric_opfunc.h:
```diff
@@ -40,7 +40,17 @@ typedef enum
+ * Maximum string size for NUMERIC output: 175 * 2 = 350
+ * = (max fractional digits (38 + 127) + 10 extra digits) * 2
+ */
+#define NUMERIC_MAX_STRING_SIZE ((DB_MAX_NUMERIC_PRECISION + DB_MAX_NUMERIC_SCALE) + 10) * 2
```
```suggestion
#define NUMERIC_MAX_STRING_SIZE  (((DB_MAX_NUMERIC_PRECISION + DB_MAX_NUMERIC_SCALE) + 10) * 2)
```
가능하면 습관적으로 괄호로 묶어 주세요.
나중에 이걸 사용할 때 의도치 않은 오류를 줄여 줍니다.

jongmin-won replied:
네, 말씀하신 내용을 반영하여 수정했습니다.
추가로, TWICE_NUM_MAX_PREC 부분도 괄호를 사용하여 보완했습니다.

ctshim commented on src/cm_common/cm_dep_tasks.c:
```diff
@@ -2272,7 +2272,7 @@ static char *
 {
 #if !defined (NUMERIC_MAX_STRING_SIZE)
-#define NUMERIC_MAX_STRING_SIZE (80 + 1)
+#define NUMERIC_MAX_STRING_SIZE (350 + 1)
```
아래처럼 하면 더 좋지 않을까요?

```suggestion
#define NUMERIC_MAX_STRING_SIZE (((DB_MAX_NUMERIC_PRECISION + DB_MAX_NUMERIC_SCALE) + 10) * 2)
```

jongmin-won replied:
네, 말씀하신 내용을 반영하여 수정했습니다.

Commit:
Fix code review feedback

kangmin5505 commented on msg/en_US.utf8/cubrid.msg:
```diff
@@ -1461,7 +1461,9 @@ Check the path of the key file (_keys) and make sure that it includes proper key
 1367 DBLINK Transaction ABORTED: %1$s
 
-1368 Last Error
+1368 Given scale of %1$d is invalid; it should be greater than %2$d and no greater than %3$d.
```
한국어와 동일한 의미를 가지려면 다음과 같이 수정이 필요할 거 같습니다.
```suggestion
1368 Given scale of %1$d is invalid; it should be between %2$d and %3$d.
```

다만 `사이`라는 표현은 경계값 포함 여부를 명확하게 나타내지 않는 거 같습니다.
따라서 다음이 더 명확할 거 같습니다.(한국어 에러 메시지도 수정 필요)
```suggestion
1368 Given scale of %1$d is invalid; it should be greater than or equal to %2$d and less than %3$d.
```

jongmin-won replied:
코드 리뷰 감사합니다.

현재 에러 메시지는 기존 ER_INVALID_PRECISION(-649) 와 통일성을 맞추기 위해 다음과 같이 작성되어 있습니다.
```
649 Given precision of %1$d is invalid; it should be greater than %2$d and no greater than %3$d.
```

말씀 주신대로 "precision 값이 0보다 크고 38보다 작거나 같아야 한다"는 의미를 가지는데, ER_INVALID_SCALE 에 그대로 적용하기에는 의미가 맞지 않는 부분이 있습니다.
이에 두 가지 방안을 검토했습니다:

1안) 에러 메시지 그대로 두고, 5번째 argument를 DB_MIN_NUMERIC_SCALE - 1로 설정
- 매번 DB_MIN_NUMERIC_SCALE 값에서 -1 연산을 해주어야 하므로, 향후 유지보수가 어려워 보임

2안) 메세지를 수정
- %2$d 부분만 "greater than or equal to" 로 바꾸고, %3$d 는 기존과 통일성을 위해 그대로 유지합니다.
```
Given scale of %1$d is invalid; it should be greater than or equal to %2$d and no greater than %3$d.
```

kangmin5505 commented on src/compat/db_macro.c:
```diff
@@ -171,18 +173,19 @@ db_value_domain_init (DB_VALUE * value, const DB_TYPE type, const int precision,
 	}
-      if (scale == DB_DEFAULT_SCALE)
+
+      value->domain.numeric_info.scale = scale;
```
160  번째 줄과 중복된 코드인지 확인 부탁드립니다.
```suggestion
```

jongmin-won replied:
네, 해당 부분은 중복된 코드로 확인됩니다.
기존 precision 관련 부분 역시 동일하게 중복된 코드로 보여 else 블록도 같이 제거했습니다.

kangmin5505 commented on src/object/object_representation.c:
```diff
@@ -2789,12 +2793,12 @@ or_put_domain (OR_BUF * buf, TP_DOMAIN * domain, int include_classoids, int is_n
 	  scale = d->scale;
-	  if (scale <= DB_DEFAULT_SCALE)
+	  if (scale < 0)
 	    {
-	      scale = 0;
+	      carrier |= (OR_DOMAIN_SCALE_MAX - 1) << OR_DOMAIN_SCALE_SHIFT;
+	      extended_scale = -scale;
 	    }
```
무엇을 의미하나요?

jongmin-won replied:
해당 부분은 domain 정보를 server에 전달하기 위한 pack 단계입니다.

extended_scale 은 pack 하려는 정보가 클 경우 사용되며,
이때 carrier 에 0xFF 를 설정하고, 이후 2970 라인에서 or_put_int() 를 통해 별도로 전달하게 됩니다.
(참고: carrier 에는 scale 정보뿐 아니라 다른 정보들도 함께 포함됩니다.)

extended_scale 을 사용한 이유는 다음과 같습니다.
- 양수/음수 scale 구분을 명확히 하기 위해
- 기존 양수 scale 로직을 그대로 적용할 경우, carrier 값에 영향을 주어 의도하지 않은 값이 전달될 수 있기 때문

kangmin5505 commented on src/object/object_representation.c:
```diff
@@ -3476,6 +3482,15 @@ unpack_domain (OR_BUF * buf, int *is_null)
 		}
+	      else if (scale == OR_DOMAIN_SCALE_MAX - 1)
+		{
+		  scale = or_get_int (buf, &rc);
+		  scale = -scale;
```
```suggestion
		  scale = -(or_get_int (buf, &rc));
```

jongmin-won replied:
네, 말씀하신 내용을 반영하여 수정했습니다.

Commit:
Fix code review feedback

Commit:
Modify error message for invalid scale range

Commit:
Removed the branching logic in Numeric Parsing that interpreted prec/scale differently depending on the total number of input digits, and modified it to consistently interpret prec in the Oracle way — from the left-most nonzero digit to the right-most significant digit.

beyondykk9 commented on src/query/numeric_opfunc.c:
```diff
@@ -3053,167 +3072,610 @@ numeric_coerce_double_to_num (double adouble, DB_C_NUMERIC num, int *prec, int *
     }
 
-  if (prec == 0 && pad_character_zero)
+  if ((int_len + frac_len) == 0)
```
(int_len == 0 && frac_len == 0)와 같은 의미인지요?

jongmin-won replied:
네 같은 의미가 맞습니다.
analyze_numeric_string()의 마지막 부분의 
```
if (!has_digit && (int_count + frac_count) == 0)
```
조건에서 int_len = 0, frac_len = 0 으로 설정하여 NULL임을 의미합니다.

따라서 아래 두 조건은 동일합니다.
1) int_len + frac_len = 0
2) int_len == 0 && frac_len == 0

kangmin5505 replied:
그렇다면 (int_len == 0 && frac_len == 0)으로 작성하는 것이 더 명확해 보입니다.

kangmin5505 replied:
그렇다면 (int_len == 0 && frac_len == 0)으로 작성하는 것이 더 명확해 보입니다.

jongmin-won replied:
코드 리뷰 감사합니다.
말씀하신 부분을 반영하여 수정했습니다.

ctshim commented:
DB_DEFAULT_SCALE이 (-1) 로 정의되어 있었던 관계로 NUMERIC의 음수 스케일 지원 정책과 혼동이 생기게 되고, 
그래서 db_value_domain_init()을 호출할 때 numeric이면 구분해서 DB_DEFAULT_NUMERIC_SCALE를 지정하고 있습니다.
전체 코드에서 scale은 `int`, 혹은 `short`으로 선언되는 것 같습니다.
DB_DEFAULT_SCALE를 (-1) 대신 (-9999) 같은 값으로 정의해 두면 코드 상 혼동이 줄어들고 더 간결한 코드가 될 것 같습니다.
한번 검토해 주세요

ctshim commented on src/compat/db_macro.c:
```diff
@@ -167,22 +169,18 @@ db_value_domain_init (DB_VALUE * value, const DB_TYPE type, const int precision,
-	  value->domain.numeric_info.scale = scale;
-	}
-      if (IS_INVALID_PRECISION (precision, DB_MAX_NUMERIC_PRECISION) || precision == 0)
+      else if (scale > (DB_MAX_NUMERIC_SCALE + DB_MAX_NUMERIC_PRECISION) || scale < DB_MIN_NUMERIC_SCALE)
```
scale의 최대값은 정해져 있는데 DB_MAX_NUMERIC_PRECISION과 더한 값과 비교해야 하나요?

jongmin-won replied:
아래 예제와 같이 (38,165) 값을 허용하기 위해 DB_MAX_NUMERIC_PRECISION과 더한 값과 비교 해야 합니다.
- cubrid (38,165)
```sql
csql> select 0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000012345678901234567890123456789012345678;

=== <Result of SELECT Command in Line 1> ===

  0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000012345678901234567890123456789012345678
======================
  0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000012345678901234567890123456789012345678
```

- oracle (38,165)
```sql
select 0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000012345678901234567890123456789012345678 FROM dual;

0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000012345678901234567890123456789012345678|
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000012345678901234567890123456789012345678|
```

ctshim replied:
scale에 대한 점검은 기존 코드처럼 유지하면 어떤 문제가 발생하나요?

hyunikn replied:
오늘 회의 시간에 얘기된대로 한다면 DB_DEFAULT_SCALE 를 더할 필요 없겠습니다.

jongmin-won replied:
NUMERIC 최대 범위 논의 이후, 다음 두 매크로를 추가하고 이를 활용해 코드를 수정했습니다.
```
#define DB_MIN_NUMERIC_SCALE -211
#define DB_MAX_NUMERIC_SCALE 252
```

hyunikn commented:
> 3. 표현 가능한 자릿수 확장
> 정수부: 최대 122자리 (Phase-1에서 유효 숫자가 38자리를 넘어갈 경우 Overflow 에러)
> 소수부: 최대 165자리 (Phase-1에서 유효 숫자가 38자리를 넘어갈 경우 39자리에서 반올림)

- 현재 큐브리드나 오라클에 없는 규칙입니다. 이 규칙이 추가된 이유가 무엇일까요.
  - 단, 오라클 reference manual에 보면 NUMBER 타입값은 절대값이 1.0 * 10 ^-130 보다 크거나 같고 1.0 * 10^126 보다 작다라고 명시되어 있습니다. 이 규칙에 의하면 오라클에서는 소수부 없이 정수만 있는 경우 자리수가 126 자로 제한됩니다. 
- 이 규칙에는 정수부도 있고, 소수부도 있는 경우에는 어떻게 된다는 것이 명확하지 않습니다. 122 + 165 = 287 자리까지 쓸 수 있는 걸까요.
- 정수부 소수부 유무에 따라 경우를 나누는 것보다 "숫자 리터럴은 소수점이나 부호 포함 254자로 제한된다" 라는 규칙 하나로 하는 것이 어떨까요.
  - 참고로, 오라클 reference manual에는 숫자 리터럴의 길이에 대한 명시는 없습니다. 
  - 단, 오라클에서 테스트를 해보니 정수부 100자리 소수부 124자리가 있는 숫자 한 건에 대해 통과를 해 주고 그 보다 커지면 에러를 냈습니다. 부호를 명시해도 마찬가지였습니다.

hyunikn commented:
> Phase-3에서 Precision 최대 40 확장 예정

40 이라는 숫자는 precision 이라고 부르지 않는 것이 좋을 것 같습니다. NUMERIC 타입의 precision 이 38 이라는 것인지 40이라는 것인지 혼동을 줍니다. 
40은 사용자가 쓴 숫자 리터럴 중에 최대 몇 자리까지 0으로 바꾸지 않고 보존할 것인가에 대한 숫자이고 NUMBER 타입의 precision과 혼동해서는 안된다고 봅니다.

hyunikn commented:
> 1-3. Numeric Parsing 중 rounding 규칙
> 소수 포함 전체 자릿수가 38자리를 초과하는 경우 -> 39자리에서 반올림하여 38자리로 저장 및 출력
> 정수 부분이 38자리를 초과하는 경우 -> 반올림 없이 Overflow 에러 발생

정수 부분에 대한 overflow 는 phase-1 임시적인 것인가요? 
결국에는 정수부 소수부 여부에 관계 없이 overflow 없이 반올림해 주어야 하겠습니다.

hyunikn commented on src/cm_common/cm_dep_tasks.c:
```diff
@@ -2272,7 +2272,7 @@ static char *
 {
 #if !defined (NUMERIC_MAX_STRING_SIZE)
-#define NUMERIC_MAX_STRING_SIZE (80 + 1)
+#define NUMERIC_MAX_STRING_SIZE (((DB_MAX_NUMERIC_PRECISION + DB_MAX_NUMERIC_SCALE) + 10) * 2)
```
(DB_MAX_NUMERIC_PRECISION + DB_MAX_NUMERIC_SCALE) 이라는 값은 어떤 의미를 갖을까요.
precision 은 유효숫자의 갯수이고 scale 은 그 자리수 위치로 반올림하거나 0을 붙여 확장하라는 의미인데, 
그 둘의 합은 어떤 의미를 갖게 될까요.
예전에 현재 큐브리드에서 부호 소수점 포함해서 숫자 리터럴의 최대 길이가 254 이다 라는 얘기를 들은 적이 있는데 그렇다면 NUMERIC_MAX_STRING_SIZE 이 254 여야 하지 않을까요.

jongmin-won replied:
현재 cm_dep_tasks의 NUMERIC_MAX_STRING_SIZE는 numeric_opfunc.h의 NUMERIC_MAX_STRING_SIZE 값과 동일하게 설정 했습니다.

NUMERIC_MAX_STRING_SIZE는 단순히 입력 리터럴 최대 길이(예: 254)가 아니라, 변환·보정·연산·출력까지 모두 담을 수 있는 최종 버퍼 크기입니다.
따라서 리터럴 길이와 동일할 필요가 없으며, TWICE_NUM_MAX_PREC의 두 배 이상으로 설정해야 안전합니다.
* 현재 CUBRID의 numeric 범위 스펙이 확정되지 않아 (DB_MAX_NUMERIC_PRECISION + DB_MAX_NUMERIC_SCALE) 값은 달라질 수 있습니다.

즉, NUMERIC_MAX_STRING_SIZE 값은 단순 입력 리터럴 길이가 아니라
1) base-256 <-> 10진 변환
2) scale 보정(0 채움)
3) 반올림 처리
4) 사칙연산 중간 결과
5) 최종 출력 버퍼
까지 고려한 최대 작업 공간을 보장하는 역할을 합니다.

hyunikn replied:
DB_MAX_NUMERIC_PRECISION 와 DB_MAX_NUMERIC_SCALE 는 더해서 어떤 의미 있는 값을 만드는 값들이 아닌 것 같습니다. 그냥 350 이라는 값을 쓰는 것이 오해의 소지가 덜 할 것 같습니다.

jongmin-won replied:
DB_MAX_NUMERIC_PRECISION과 DB_MAX_NUMERIC_SCALE은 단순히 더해 값을 만드는 매크로가 아닙니다.
이를 더하는 이유는 문자열 표현 시 필요한 최대 자릿수를 스펙 기반으로 산정하기 위함입니다.
- 표현 가능한 최댓값 기준: PRECISION(43) + |MIN_SCALE|(211) = 254
- 버퍼를 2의 배수 정렬로 맞추기 위해 +2 -> 256으로 보정
- 스케일 정렬 과정에서의 일시적 자릿수 증가를 흡수하기 위해 * 2 → 512로 확정
```
NUMERIC_MAX_STRING_SIZE = (43 + 211 + 2) * 2 = 512
```

따라서 질문 주신 것처럼 단순히 "350" 같은 상수를 사용하면 계산 근거가 불명확해지고, "왜 350이지?"라는 의문을 줄 수 있습니다.
이에 따라 위 계산식을 바탕으로 NUMERIC_MAX_STRING_SIZE 값을 정의했습니다.

hyunikn commented on src/compat/dbtype_def.h:
```diff
@@ -758,7 +762,8 @@ extern "C"
       unsigned char type;
       unsigned char precision;
-      unsigned char scale;
+      int scale;
```
int 말고 char 로도 -84 ~ 127 범위를 커버할 수 있습니다. char 는 어떤가요

hyunikn replied:
회의 시간에 왜 int 인지 알게 되었습니다.

hyunikn commented on src/compat/dbtype_def.h:
```diff
@@ -758,7 +762,8 @@ extern "C"
       unsigned char precision;
-      unsigned char scale;
+      int scale;
+      bool has_round;
```
has_round 는 어떤 의미인지 주석이 있으면 좋을 것 같습니다.

jongmin-won replied:
has_round 변수명은 임의로 작성된 것으로, already_rounded로 변경할 예정입니다.
말씀 주신 대로 간단한 설명 주석도 함께 추가하겠습니다.

보다 상세한 내용은 numeric_db_value_coerce_to_num()의 3921~3931 라인에 주석으로 기재해 두었습니다.

hyunikn replied:
필요한 반올림을 이미 했으므로 다시 할 필요 없음 을 표시하는 flag 인가요

jongmin-won replied:
네, Numeric Parsing 단계에서 이미 반올림이 수행되었기 때문에 이후 로직에서 다시 반올림되는 것을 방지하고자 flag를 추가했었습니다.
그러나 구현 오류로 인해 문제가 발생하는 것으로 확인되어, 해당 flag는 제거했습니다.

ctshim commented on src/object/object_representation.c:
```diff
@@ -2621,6 +2621,10 @@ or_packed_domain_size (TP_DOMAIN * domain, int include_classoids)
 	    {
 	      precision = DB_MAX_NUMERIC_PRECISION;
 	    }
+	  if (scale < 0)
+	    {
+	      size += OR_INT_SIZE;
+	    }
```
scale이 음수일 때만 전송하는 데이터가 달라지는 것인가요?

jongmin-won replied:
음수 값의 경우 carrier에 scale 정보를 담지 않고 or_put_int()을 사용하여 전달 하려 했습니다.

하지만 이후 논의 결과, scale 값이 128 이상이거나 음수일 경우 extended_scale을 사용하도록 수정했습니다.

수정 내용
- carrier에 저장 가능한 scale 범위 : 0 ~ 127
- extended_scale에 저장하는 scale 범위 :
  - scale >= 128
  - -1 ~ -211

hyunikn commented on src/compat/db_macro.c:
```diff
@@ -167,22 +169,18 @@ db_value_domain_init (DB_VALUE * value, const DB_TYPE type, const int precision,
-	{
-	  value->domain.numeric_info.precision = precision;
-	}
-      if (scale == DB_DEFAULT_SCALE)
```
scale 이 DB_DEFAULT_SCALE 일 때 처리 로직이 왜 빠졌을까요.

hyunikn replied:
아래 답변을 보니까 이해가 됩니다. 
scale 이 -1 값을 가질 수 있게 되어서 그런 것 이군요.
그렇다면 DB_DEFAULT_SCALE 의 값을 scale 이 가질 수 없는 -85 이하의 값으로 바꾸면 어떨까요.
그렇게 하면 이후 db_value_domain_init 호출할 때 DB_DEFAULT_SCALE 대신 DB_DEFAULT_NUMERIC_SCALE 로 바꾼 것을 원래 코드 그대로 쓸 수 있을 것 같습니다.

ctshim replied:
저도 이와 관련해서 의견 드린적이 있는데, 한번 반영 검토해 주세요.
이것만 반영되어도 변경된 코드가 줄어 드는 곳이 많을 듯 합니다.^^

jongmin-won replied:
리뷰 의견을 반영하여 DB_DEFAULT_SCALE 값을 -1에서 -9999로 변경하였으며,
이전에 제거했던 DB_DEFAULT_SCALE 관련 조건들도 모두 원복했습니다.

hyunikn commented on src/executables/csql_result_format.c:
```diff
@@ -914,6 +914,18 @@ numeric_to_string (DB_VALUE * value, bool commas)
    */
   prec = DB_VALUE_PRECISION (value);
+  scale = DB_VALUE_SCALE (value);
+  // This condition was added due to the if (strlen (str_buf) > max_length - 1) condition.
```
주석에서 말하는 condition 은 어디에 있는 condition 인가요.

jongmin-won replied:
938 라인의 `if (strlen (str_buf) > max_length - 1)` 조건 입니다.

hyunikn replied:
문장 끝의 condition 말고 "This condition" 이 무엇인가 하는 질문입니다.

jongmin-won replied:
네 말씀 주신 condition은 920라인의 `if (scale < 0)` 및 924라인의 `else if (scale > prec)` 조건을 의미합니다.

hyunikn commented on src/executables/csql_result_format.c:
```diff
@@ -914,6 +914,18 @@ numeric_to_string (DB_VALUE * value, bool commas)
+  // If removed, NUM OVERFLOW will be output.
+  if (scale < 0)
+    {
+      prec -= scale;
```
precision 이 변하는 것은 아니니 가독성을 위해 새로운 변수를 도입하는 것이 어떨까요.
이름 짓자면 digits 같은.

jongmin-won replied:
930, 931 라인에서 return_string의 길이는 prec + comma_length + 3으로 계산되고 있습니다.
precision 자체가 변하는 것은 아니므로, 말씀하신 대로 digits와 같은 별도 변수를 도입해 max_length를 계산하는 방식이 가독성 측면에서 더 적절해 보입니다.
```
int digits = (scale < 0) ? (prec - scale) :
             (scale > prec ? (prec + scale) : prec);

comma_length = COMMAS_OFFSET (commas, digits);
max_length = digits + comma_length + 3;
```
이렇게 하면 원래의 prec은 그대로 두면서, 출력 길이 계산에 필요한 자릿수(digits)만 별도로 관리할 수 있게 수정하겠습니다.

hyunikn replied:
그런데, scale > prec 인 경우는 digits 는 prec + scale 이 아니라 scale 아닐까요?
예: 0.001234 : numeric(4, 6) 에서 digits 001234의 길이는 10이 아니라 6.

jongmin-won replied:
네 의견 주신 대로 prec + scale이 아닌,  scale로 수정 했습니다.

hyunikn commented on src/executables/csql_result_format.c:
```diff
@@ -914,6 +914,18 @@ numeric_to_string (DB_VALUE * value, bool commas)
+    }
+  else if (scale > prec)
+    {
+      prec += scale;
```
(질문) 이 경우 소수점 오른쪽 숫자 갯수 scale 과 맨 왼쪽 '0.' 의 길이를 합하여 scale + 2 가 좀 더 정확한 값이겠지요?

jongmin-won replied:
`numeric_db_value_print()` 내부에서 이미 '0.' 부분의 길이를 별도로 처리하고 있으므로, 계산 시 scale + 2를 추가로 고려할 필요는 없습니다.

hyunikn commented on src/object/object_domain.c:
```diff
@@ -3391,11 +3391,6 @@ tp_domain_resolve_value (const DB_VALUE * val, TP_DOMAIN * dbuf)
 	      domain->precision = DB_DEFAULT_NUMERIC_PRECISION;
 	    }
 
-	  if (domain->scale == -1)
```
이 부분은 왜 삭제되었을까요.

jongmin-won replied:
NUMERIC 타입의 scale에 음수 표현이 가능하여,  `domain->scale == -1` 인 경우  `DB_DEFAULT_NUMERIC_SCALE(0)` 로 변경 되어 삭제 하였습니다.

이전 comment에서 말씀 주신 것 과 같이 `DB_DEFAULT_SCALE(-1)` 을  다른 값(ex. -9999 등)으로 변경하면,  아래 예제와 같이 수정되어 추가될 예정 입니다. 

```
if (domain->scale == DB_DEFAULT_SCALE) 
...
```

hyunikn commented on src/object/object_representation.c:
```diff
@@ -2621,6 +2621,10 @@ or_packed_domain_size (TP_DOMAIN * domain, int include_classoids)
 	    {
 	      precision = DB_MAX_NUMERIC_PRECISION;
 	    }
+	  if (scale < 0)
```
(질문) scale 이 음수이면 domain 을 pack 한 결과가 int size 만큼 커지나요?

jongmin-won replied:
음수 값의 경우 carrier에 scale 정보를 담지 않고, extended_scale와 동일하게 or_put_int()을 사용해 전달하려 했습니다.
이 경우 or_put_int()를 쓰면 or_packed_domain_size()에서 size를 int 크기만큼 늘려줘야 합니다.

hyunikn commented on src/object/object_representation.c:
```diff
@@ -2789,12 +2793,12 @@ or_put_domain (OR_BUF * buf, TP_DOMAIN * domain, int include_classoids, int is_n
 	    }
-
-	  if (scale < OR_DOMAIN_SCALE_MAX)
+	  else if (scale < OR_DOMAIN_SCALE_MAX)
```
2805 line 의 else block 은 도달하지 않으니 assert (false) 로 바꾸거나 지우는 것이 어떨까요.

jongmin-won replied:
9/15일 춘택님과 오프라인으로 논의 한 결과, 
scale 값이 "128 이상"이거나 "음수"일 경우 extended_scale을 사용하도록 수정했습니다.

수정 내용
- carrier에 저장 가능한 scale 범위 : 0 ~ 127
- extended_scale에 저장하는 scale 범위 :
  - scale >= 128
  - -1 ~ -211

hyunikn commented on src/object/object_representation.c:
```diff
@@ -3476,6 +3482,14 @@ unpack_domain (OR_BUF * buf, int *is_null)
 		      goto error;
 		    }
 		}
+	      else if (scale == OR_DOMAIN_SCALE_MAX - 1)
```
. OR_DOMAIN_SCALE_MAX - 1 의 값은 254 인데, numeric domain 의 scale 이 254와 같을 수 없을 것 같습니다. 
. 3477 라인에서 255 와 같을 수도 없을 것 같습니다. - 가능하면 지우거나 assert (false) 를 두는 것이 좋을 것 같습니다.

jongmin-won replied:
기존에는 OR_DOMAIN_SCALE_MAX(0xFF) 처리 방식과 동일하게, 0xFE(254) 값에 담긴 값을 or_get_int()로 읽어 음수 scale을 설정하려 했습니다.

하지만 이후 논의 결과, scale 값이 "128 이상"이거나 "음수"일 경우 extended_scale을 사용하도록 수정했습니다.

수정 내용
- carrier에 저장 가능한 scale 범위 : 0 ~ 127
- extended_scale에 저장하는 scale 범위 :
  - scale >= 128
  - -1 ~ -211

ctshim commented on src/parser/csql_grammar.y:
```diff
@@ -23583,6 +23582,31 @@ unsigned_integer
+			PARSER_SAVE_ERR_CONTEXT ($$, @$.buffer_pos)
+
+		DBG_PRINT}}
+	| '-' UNSIGNED_INTEGER
```
아래와 같은 방식으로 대체하면 더 좋겠습니다. 이에 따라 하위 코드도 좀 다르겠구요
```suggestion
	| '-' unsigned_int32
```

hyunikn replied:
@ctshim unsigned_int32에 어떤 장점이 있을까요?

jongmin-won replied:
테이블 생성 시 NUMERIC 컬럼의 scale 값을 입력할 때, 키워드 변경으로 인해 에러 메시지가 달라집니다.

AS-IS (UNSIGNED_INTEGER 유지)
```sql
create table t1 (col1 numeric(1,-99999999999999999999999999999999999999)); 

In line 1, column 72,
ERROR: Numeric scale specifier is out of range -84 to 127
```

TO-BE (UNSIGNED_INTEGER -> unsigned_int32 변경)
```sql
create table t1 (col1 numeric(1,-99999999999999999999999999999999999999)); 

In line 1, column 72,
ERROR: Invalid unsigned integer 99999999999999999999999999999999999999.
```

jongmin-won replied:
다른 키워드와의 통일성을 위해, `opt_prec_2`에서 사용 중인 `unsigned_integer`로 수정 했습니다.

ctshim commented on src/parser/csql_grammar.y:
```diff
@@ -23583,6 +23582,31 @@ unsigned_integer
 	;
 
+numeric_scale_integer
+	: unsigned_integer
```
```suggestion
	: unsigned_int32
```

jongmin-won replied:
위 comment와 같은 이유로 `unsigned_int32`로 변경하지 않고, 기존 `unsigned_integer`를 유지하겠습니다.

AS-IS (unsigned_integer 유지)
```sql
create table t1 (col1 numeric(1,99999999999999999999999999999999999999)); 

In line 1, column 71,
ERROR: Numeric scale specifier is out of range -84 to 127
```

TO-BE (unsigned_integer -> unsigned_int32 변경)
```sql
create table t1 (col1 numeric(1,99999999999999999999999999999999999999)); 

In line 1, column 71,
ERROR: Invalid unsigned integer 99999999999999999999999999999999999999.
```

ctshim commented on src/parser/semantic_check.c:
```diff
@@ -2293,10 +2289,6 @@ pt_union_compatible (PARSER_CONTEXT * parser, PT_NODE * item1, PT_NODE * item2,
 	    {
 	      data_type->info.data_type.dec_precision =
 		(DB_MAX_NUMERIC_PRECISION - data_type->info.data_type.dec_precision);
-	      if (data_type->info.data_type.dec_precision < 0)
-		{
-		  data_type->info.data_type.dec_precision = 0;
-		}
```
기존에 dec_precision 이 `0`인 경우는 어떤 의미였는지 궁금하고,
현재 코드에서는 왜 삭제한 것인지도 궁금합니다.

jongmin-won replied:
data_type의 dec_precision은 db_value의 scale과 동일합니다.
따라서 현재 코드에서 해당 부분을 삭제한 이유는 음수 scale을 표현하기 위함입니다.

ctshim commented on src/query/numeric_opfunc.c:
```diff
@@ -3053,167 +3072,610 @@ numeric_coerce_double_to_num (double adouble, DB_C_NUMERIC num, int *prec, int *
+	      if (astring[parse_pos] == '-')
 		{
-		  break;
+		  *negate_value = true;
+		  parse_pos++;
```
3146 라인의 '}'는 3143 라인 바로 이후로 옮겨 져야 합니다.
부호가 '+'일 때  오동작합니다.

jongmin-won replied:
아래와 같이 테스트한 결과 정상적으로 출력되는 것을 확인했습니다.
```sql
csql> select +123.0;

=== <Result of SELECT Command in Line 1> ===

  123.0               
======================
  123.0    

csql> select -123.0;

=== <Result of SELECT Command in Line 1> ===

  -123.0              
======================
  -123.0  
```

numeric_coerce_string_to_num(astring, …) 함수의 첫 번째 인자 astring 값을 확인해 보니, 
사용자가 입력한 "+123.0"이 아닌 "123.0"으로 전달되고 있었습니다.

사용자가 입력한 '+' 기호가 어디에서 제거되는지는 추가로 확인해 보겠습니다.

jongmin-won replied:
String을 Numeric으로 캐스팅할 때, 말씀하신 대로 수정하지 않으면 오류가 발생하는 것을 확인하여 수정 했습니다.
`SELECT cast('+12000.123456' as numeric);`

ctshim commented on src/query/numeric_opfunc.c:
```diff
@@ -3241,7 +3703,7 @@ numeric_coerce_num_to_num (DB_C_NUMERIC src_num, int src_prec, int src_scale, in
 {
   int ret = NO_ERROR;
-  char num_string[DB_MAX_NUMERIC_PRECISION * 4];
+  char num_string[NUMERIC_MAX_STRING_SIZE];
```
다른 수정된 위치는 `NUMERIC_MAX_STRING_SIZE + 1`을 사용하던데 여기는 `+ 1`이 필요 없을까요?

jongmin-won replied:
다른 위치에서 사용된 `NUMERIC_MAX_STRING_SIZE + 1`은 큰 의미가 없습니다.
이미 `NUMERIC_MAX_STRING_SIZE` 값이 `TWICE_NUM_MAX_PREC`의 두 배로 설정되어 있어 충분한 여유가 있다고 판단됩니다.

가독성과 통일성을 위해 필요하다면, 다른 부분의 `NUMERIC_MAX_STRING_SIZE + 1` 역시 `NUMERIC_MAX_STRING_SIZE`로 수정 했습니다.

ctshim commented on src/query/numeric_opfunc.c:
```diff
@@ -3847,7 +4320,7 @@ numeric_db_value_coerce_from_num_strict (DB_VALUE * src, DB_VALUE * dest)
 numeric_db_value_print (const DB_VALUE * val, char *buf)
 {
-  char temp[80];
+  char temp[NUMERIC_MAX_STRING_SIZE];
```
다른 수정된 위치는 `NUMERIC_MAX_STRING_SIZE + 1`을 사용하던데 여기는 `+ 1`이 필요 없을까요?

jongmin-won replied:
다른 위치에서 사용된 `NUMERIC_MAX_STRING_SIZE + 1`은 큰 의미가 없습니다.
이미 `NUMERIC_MAX_STRING_SIZE` 값이 `TWICE_NUM_MAX_PREC`의 두 배로 설정되어 있어 충분한 여유가 있다고 판단됩니다.

가독성과 통일성을 위해 필요하다면, 다른 부분의 `NUMERIC_MAX_STRING_SIZE + 1` 역시 `NUMERIC_MAX_STRING_SIZE`로 수정 했습니다.

ctshim commented on src/query/numeric_opfunc.c:
```diff
@@ -3923,6 +4396,30 @@ numeric_db_value_print (const DB_VALUE * val, char *buf)
+      if (!found_first_non_zero)
+	{
+	  // If no digit found: output only a single '0'
+	  buf[0] = '0';
+	  buf[1] = '\0';
+	  nbuf = 1;
```
여기는 에러로 처리해야 하는 상황이 아닌가요?
`0` 이 아니고 차라리 NULL로 처리되어야 할 것은 아닌지?

jongmin-won replied:
해당 부분을 테스트할 수 있는 예제는 `select cast(0 as numeric(1,-1));` 가 있습니다.

만약, 말씀 하신 코드를 제거하고 테스트 한 결과는 아래와 같습니다. (NULL, ERR는 아직 추가하지 않음)
```sql
csql> select cast(0 as numeric(1,-1));

=== <Result of SELECT Command in Line 1> ===

   cast(0 as numeric(1,-1))
======================
                      
```

하지만, 관련 Case를 오라클에서 테스트 결과 0이 나와,  동일하게 처리하기 위해 추가한 부분 입니다.
```sql
-- 1) Oracle 테스트
select cast(0 as number(1,-1));
CAST(0ASNUMBER(1,-1))|
---------------------+
                    0|

-- 2) 해당 코드를 포함하여 CUBRID 테스트
csql> select cast(0 as numeric(1,-1));

=== <Result of SELECT Command in Line 1> ===

   cast(0 as numeric(1,-1))
======================
  0  
```

hyunikn commented on src/query/numeric_opfunc.c:
```diff
@@ -3456,6 +3918,17 @@ numeric_db_value_coerce_to_num (DB_VALUE * src, DB_VALUE * dest, DB_DATA_STATUS
+
+	/*
+	 * When the domain has a negative scale and the input value contains a fractional part
+	 * e.g.) numeric(38, -1), insert (1234567890...6789.0)
```
주석의 예에서 scale = 1, desired_scale = -1 인 상황인가요?

jongmin-won replied:
결론부터 말씀드리면, 구현 상 오류로 인해 발생한 문제이며, 해당 예제와 has_round 플래그는 제거할 예정입니다.
numeric_coerce_string_to_num() -> determine_prec_scale() 처리 중 특정 케이스에서 의도하지 않은 출력이 발생하여 has_round 플래그를 추가했으나, 원인은 구현 상 오류가 있었던 것으로 확인되었습니다.

- 특정 케이스
```
CREATE TABLE t1 (col1 NUMERIC(38, -1));
INSERT INTO t1 VALUES (123456789012345678901234567890123456789.0);
SELECT * FROM t1;

-- has_round 미사용 시 : 12345678901234567890123456789012345680 (버그)
-- has_round 사용 시   : 123456789012345678901234567890123456790
```

질문 주신 내용으로 다시 돌아와 설명드리겠습니다.

123456789012345678901234567890123456789.0 값은 numeric_coerce_string_to_num() 에서 다음과 같이 DB_VALUE로 생성됩니다.
- prec : 38
- scale : 0 (버그, 정상 case는 -1로 설정되어야 함)
- data : 12345678901234567890123456789012345679

따라서 scale = 0, desired_scale = -1이 됩니다.

jongmin-won replied:
유효 숫자 범위를 초과할 경우, 초과된 부분은 음수 scale로 처리됩니다.

Oracle 테스트 결과, 입력값이 "정수 + 소수" 형태의 경우 정수부가 유효 숫자 범위를 넘어가면 소수부의 scale은 무시되고, 
정수부의 초과 자릿수만큼 음수 scale이 설정됩니다.

- Oracle 테스트 결과
- 입력 :
  - 전체 자릿수 : 54
  - 정수 자릿수 : 50
  - 소수 자릿수 : 4
- 결과 :
  - 전체 자릿수 : 50
  - 유효 숫자 : 39
  - 0 개수 : 11
```
select 12345678901234567890123456789012345678901234567890.1234 from dual;

12345678901234567890123456789012345678901234567890.1234|
-------------------------------------------------------+
     12345678901234567890123456789012345678900000000000|
```

hyunikn commented on src/query/numeric_opfunc.c:
```diff
@@ -3456,6 +3918,17 @@ numeric_db_value_coerce_to_num (DB_VALUE * src, DB_VALUE * dest, DB_DATA_STATUS
+	 * In this case, numeric_coerce_string_to_num already rounds the value from the leftmost 39 digits, resulting in 38 digits.
+	 * Therefore, a flag was added to skip redundant rounding in numeric_coerce_num_to_num.
+	 */
+	if (src->domain.numeric_info.has_round && desired_scale < 0)
```
주석의 예로 얘기하자면, 
이 라인에 도달하기 전 numeric_coerce_string_to_num() 에서 src 가 scale -1 에 해당하는 자리수 (십의 자리) 로 이미 반올림 되었을 것이므로 scale 을 desired_scale 인 -1 로 갱신해도 좋다. 이런 의미일까요.
이 if 문과 scale = desired_scale; assign 문에 대해 잘 이해가 가지 않습니다.

jongmin-won replied:
앞서 설명드린 특정 케이스를 이어서 말씀드리겠습니다.

`if (src->domain.numeric_info.has_round && desired_scale < 0)` 해당 조건의 의미는,
numeric_coerce_string_to_num() 단계에서 유효 숫자가 범위를 초과하여 이미 반올림이 수행됐고 desired_scale이 음수인 경우, desired_scale 값이 scale에 그대로 전달됩니다.

이후 numeric_coerce_num_to_num() 함수 내부에서 다음 조건에 만족하여 넘어가게 됩니다.
```
if (src_prec <= dest_prec && src_scale == dest_scale)
```

만약 scale = 0, desired_scale = -1인 경우에는 위 조건을 만족하지 못하므로, 이후 단계에서 다시 반올림이 수행됩니다.

예를 들어 DB_VALUE dest 데이터가 다음과 같을 때
- 반올림 전 DB_VALUE dest 값
```
data : 12345678901234567890123456789012345679
```

이 상태에서 scale = -1 기준으로 반올림을 적용하면 최종 값은 아래와 같이 처리 됩니다.
```
data : 1234567890123456789012345678901234568
```

Commit:
1. Extended maximum range of NUMERIC
2. Three modifications in NUMERIC parsing
- Apply rounding even when only the integer part exists
- When both integer and fractional parts exist, handle as negative scale if the integer significant digits exceed the limit
- Simplify parsing of the integer part (e.g., 123000 is treated as prec=6; previously it was split into 3 significant digits + 3 trailing zeros) Note: the fractional part is still calculated separately
3. Incorporated code review feedback

beyondykk9 commented on src/base/system_parameter.c:
```diff
@@ -8856,13 +8856,19 @@ sysprm_generate_new_value (SYSPRM_PARAM * prm, const char *value, bool check, SY
 	if (prm->id == PRM_ID_STORED_PROCEDURE_RETURN_NUMERIC_SIZE)
 	  {
 	    /*  
+	     * The maximum allowable range of NUMERIC values is:
```
Floating Point Numeric이 도입되면, 이 시스템 파라미터는 deprecate될 예정입니다.
따라서, 지금 수정할 필요는 없습니다.

jongmin-won replied:
네, 해당 부분은 향후 Floating-Point Numeric 도입 시점에 수정하겠습니다.
지금은 다시 원복해 두겠습니다.

beyondykk9 commented on src/compat/dbtype_def.h:
```diff
@@ -539,14 +539,25 @@ extern "C"
 #define DB_MAX_BIT_LENGTH 0x3fffffff
 
+/* The maximum precision that can be specified for a numeric domain in SQL grammar. */
+#define DB_GRAMMAR_MAX_NUMERIC_PRECISION 38
```
DB_MAX_FIXED_NUMERIC_PRECISION이 가독성이 더 좋지 않을까요?
마찬가지로, DB_MIN_FIXED_NUMERIC_SCALE, DB_MAX_FIXED_NUMERIC_SCALE이 어떨지요?

jongmin-won replied:
네, 말씀하신 내용을 반영하여 수정했습니다.

Commit:
Fix code review feedback

ctshim commented on src/object/object_representation.c:
```diff
@@ -2690,6 +2700,11 @@ or_packed_domain_size (TP_DOMAIN * domain, int include_classoids)
 	  size += OR_INT_SIZE;
 	}
 
+      if (scale < 0)
```
```suggestion
     else if (scale < 0)
```

jongmin-won replied:
어제 오프라인에서 논의한 내용대로, 
scale 값이 "128 이상"이거나 "음수"일 경우 extended_scale을 사용하도록 수정했습니다.

수정 내용
- carrier에 저장 가능한 scale 범위 : 0 ~ 127
- extended_scale에 저장하는 scale 범위 :
  - scale >= 128
  - -1 ~ -211

따라서, (scale < 0) 로 비교하는 조건은 제거 됐습니다.

ctshim commented on src/object/object_representation.c:
```diff
@@ -2794,7 +2810,12 @@ or_put_domain (OR_BUF * buf, TP_DOMAIN * domain, int include_classoids, int is_n
 	    }
 
-	  if (scale < OR_DOMAIN_SCALE_MAX)
+	  if (scale < 0)
```
2808라인에서 `if (scale == DB_DEFAULT_SCALE)` 로 물어 보지 않고 부등호를 넣은 이유가 혹시 모를 쓰레기 값에 대해서도 0으로 처리해 주려고 했던 것이었을까요?
원래 코드의 의도를 한번 고민해 봐야겠습니다.  기존에는 -1이었었기에 유효한 값이 아닌 첫번째 값을 가지고 있었는데, DB_DEFAULT_SCALE가 좀 유효한 값과 거리가 있으면 않되는 것일지에 대해서요

jongmin-won replied:
현재 develop에서는 scale이 음수가 될 수 없으므로 if (scale <= DB_DEFAULT_SCALE) 조건을 사용했던 것으로 보입니다.
DB_DEFAULT_SCALE(-1)은 domain_init 단계에서 주로 사용되며, 가끔 넘어올 수 있는 값이라고 생각해 `<=` 비교를 했던 것 같습니다.
하지만 실제로는 `==`가 맞는 조건이고, 왜 `<=`를 사용했는지는 정확히 추측하기 어렵습니다.

만약, develop 에서 scale의 쓰레기 값(음수)이 들어오는 경우까지 0으로 처리하려는 의도가 있었다면,
지금은 DB_DEFAULT_SCALE(-9999) 대신 DB_MIN_NUMERIC_SCALE(-211)을 사용하고, 조건을 아래와 같이 수정하면 동일한 기대 결과를 얻을 수 있다고 판단됩니다.
```
if (scale < DB_MIN_NUMERIC_SCALE)
  {
    scale = 0;
  }
```

ctshim commented on src/object/object_representation.c:
```diff
@@ -2556,7 +2556,17 @@ or_decode (const char *buffer, char *dest, int size)
+ * Note: Do not assign values beyond 253 directly.
+ *       Use 0xFE or 0xFF encoding for negative or extended scales.
+ */
+#define OR_DOMAIN_SCALE_NEG_FLAG       (0xFE)	/* negative scale follows */
+#define OR_DOMAIN_SCALE_MAX            (0xFF)	/* extended scale follows */
```
통상 carrier에 담기는 FLAG는 한 비트만 설정해서 사용하고 있습니다.
여기서 0xFE를 사용해야 하는 이유가 있나요?
가령 (0x800) 을 사용하지 않고 있는 이유요.

jongmin-won replied:
어제 오프라인에서 논의한 내용대로, 
scale 값이 "128 이상"이거나 "음수"일 경우 extended_scale을 사용하도록 수정했습니다.

수정 내용
- carrier에 저장 가능한 scale 범위 : 0 ~ 127
- extended_scale에 저장하는 scale 범위 :
  - scale >= 128
  - -1 ~ -211

ctshim commented on src/object/object_representation.c:
```diff
@@ -3227,6 +3257,11 @@ unpack_domain_2 (OR_BUF * buf, int *is_null)
 	      scale = or_get_int (buf, &rc);
 	    }
 
+	  if (scale == OR_DOMAIN_SCALE_NEG_FLAG)
```
3257 라인에서 얻은 `scale`이 OR_DOMAIN_SCALE_NEG_FLAG와 같아 질 수 없다면 `else if`문을 사용하는게 좋겠습니다.

jongmin-won replied:
어제 오프라인에서 논의한 내용대로, 
scale 값이 "128 이상"이거나 "음수"일 경우 extended_scale을 사용하도록 수정했습니다.

수정 내용
- carrier에 저장 가능한 scale 범위 : 0 ~ 127
- extended_scale에 저장하는 scale 범위 :
  - scale >= 128
  - -1 ~ -211

따라서, `if (scale == OR_DOMAIN_SCALE_NEG_FLAG)` 해당 조건은 제거 됐습니다.

ctshim commented on src/object/object_representation.c:
```diff
@@ -3476,6 +3513,15 @@ unpack_domain (OR_BUF * buf, int *is_null)
+
+	      if (scale == OR_DOMAIN_SCALE_NEG_FLAG)
+		{
+		  scale = -(or_get_int (buf, &rc));
```
보낼 때 굳이 양수로 만들어서 보내고 받은 후에 다시 음수로 전환하는 이유가 따로 있나요?
불필요한 동작 같아 보입니다.

추가로 OR_DOMAIN_SCALE_MAX와 OR_DOMAIN_SCALE_NEG_FLAG가 본질적으로 다르지 않다고 생각 됩니다.
이 두 가지 모두 carrier 에서 scale을 얻지 못하고 별도의 int 값이 뒤 따라 온다는 공통점이 있습니다.
즉 carrier에 추가된 공간에서 scale을 얻어라라는 bit를 하나 세우면 해결 될 것 같은데 로직을 복잡하게 구분하고 있는 듯 합니다.

jongmin-won replied:
OR_DOMAIN_SCALE_NEG_FLAG는 제거되었으며, 
extended_scale에 값을 담을 때는 양수로 전달하고, 수신 측에서 음수로 변환하지 않고 scale 값 그대로 전달 하도록 수정했습니다.

ctshim commented on src/query/numeric_opfunc.c:
```diff
@@ -2675,7 +2701,7 @@ numeric_get_integral_part (const DB_C_NUMERIC num, const int src_prec, const int
 
   numeric_zero (dest, DB_NUMERIC_BUF_SIZE);
-  memset (new_dec_num, 0, DB_MAX_NUMERIC_PRECISION + 1);
+  memset (new_dec_num, 0, (DB_MAX_NUMERIC_PRECISION - DB_MIN_NUMERIC_SCALE) + 1);
```
`(DB_MAX_NUMERIC_PRECISION - DB_MIN_NUMERIC_SCALE) + 2` 크기의 버퍼인데 `(DB_MAX_NUMERIC_PRECISION - DB_MIN_NUMERIC_SCALE) + 1` 크기만 meset해야 할 이유가 있나요?
특별한 이유가 없다면 아래처럼 해 주는게 코드 이해가 쉬울 듯 합니다.

`memset (new_dec_num, 0, sizeof(new_dec_num));`

jongmin-won replied:
NUMERIC 범위 스펙에 맞게 수정하는 과정에서 누락되었던 부분입니다.  
말씀해주신 대로 `sizeof(new_dec_num)`로 수정했습니다.

ctshim commented on src/query/numeric_opfunc.c:
```diff
@@ -3053,161 +3079,556 @@ numeric_coerce_double_to_num (double adouble, DB_C_NUMERIC num, int *prec, int *
     {
-      skip_size = 1;
-      if (astring[i] == '.')
+      skip = 1;
```
skip을 초기화 하는 작업은 필요 없어 보입니다.
intl_is_space() 내부에서 다시 초기화를 하네요

jongmin-won replied:
코드 리뷰 감사합니다.
말씀하신 부분을 반영하여 수정했습니다.

ctshim commented on src/query/numeric_opfunc.c:
```diff
@@ -3053,161 +3079,556 @@ numeric_coerce_double_to_num (double adouble, DB_C_NUMERIC num, int *prec, int *
+	  return DOMAIN_INCOMPATIBLE;
+	}
+    }
+
```
'0' 만 있는 경우 has_digit가 설정되지 않게 됩니다.
문제의 소지가 없는지요?
추가로 "+0", "-0"과 같은 경우 어떻게 처리 되어야 할 지에 대해서도 검토 필요해 보입니다.

jongmin-won replied:
아래 comment 중 while 루프에서 has_digit을 반복적으로 확인하지 않도록 수정했으며,
has_digit의 위치도 조정하여 다음 3가지 케이스가 모두 정상 동작합니다.

```sql
select cast('0' as numeric);
select cast('+0' as numeric);
select cast('-0' as numeric);
```
참고로 단순한 `select 0`은 numeric_coerce_string_to_num() 함수로 들어오지 않습니다.

has_digit 위치 수정 이유
1. 의미 명확화(정의 일치)
- has_digit는 "숫자를 하나라도 읽었는가"를 나타내는 플래그입니다. 
  - 이는 '유효 숫자' 여부가 아니라, 단순히 입력에 숫자가 존재하는지를 확인하여 NULL 여부를 판별하기 위한 것입니다.
- 기존에는 '.'에서 has_digit = true를 세팅해 의미가 흐려졌습니다. 
- 이제 숫자(0–9)를 읽은 경우에만 true로 설정합니다.
2. 버그 수정(‘.’ 처리)
- 입력이 `'. '`인 경우 결과값은 NULL로 출력됩니다. ('.'에서 has_digit = true로 설정해도 결과는 동일)
- 그러나 기존 로직에서는 analyze_numeric_string() 단계가 아닌 다른 경로에서 NULL 처리가 이루어졌습니다.
- 이를 수정하여 analyze_numeric_string() 단계에서 일관되게 NULL 판정이 이루어지도록 했습니다.
```sql
select cast('.    ' as numeric);
```

ctshim commented on src/query/numeric_opfunc.c:
```diff
@@ -3053,161 +3079,556 @@ numeric_coerce_double_to_num (double adouble, DB_C_NUMERIC num, int *prec, int *
-	      /* Characters excluding digit, space and comma are not acceptable. */
-	      ret = DOMAIN_INCOMPATIBLE;
+	      // Integer with decimal: 99999.99999... -> 100000.0
+	      memmove (out_str + 1, out_str, phase_1_tmp_max_prec);
```
`memset(out_str + 1, '0', phase_1_tmp_max_prec);`으로 하는 것이 어떤가요?
미미하겠지만 속도 차이가 있을 수 있습니다.

jongmin-won replied:
코드 리뷰 감사합니다.
말씀하신 부분을 반영하여 수정했습니다.

ctshim commented on src/query/numeric_opfunc.c:
```diff
@@ -3923,6 +4344,30 @@ numeric_db_value_print (const DB_VALUE * val, char *buf)
+	{
+	  // If no digit found: output only a single '0'
+	  buf[0] = '0';
+	  buf[1] = '\0';
```
4368라인에서 처리하므로 4357라인은 필요 없는 내용입니다.

jongmin-won replied:
코드 리뷰 감사합니다.
말씀하신 부분을 반영하여 수정했습니다.

kangmin5505 commented on src/compat/db_macro.c:
```diff
@@ -3058,7 +3067,8 @@ db_value_is_corrupted (const DB_VALUE * value)
     case DB_TYPE_NUMERIC:
-      if (IS_INVALID_PRECISION (value->domain.numeric_info.precision, DB_MAX_NUMERIC_PRECISION))
+      if (IS_INVALID_PRECISION (value->domain.numeric_info.precision, DB_MAX_NUMERIC_PRECISION)
+	  && IS_INVALID_NUMERIC_SCALE (value->domain.numeric_info.scale, DB_MIN_NUMERIC_SCALE, DB_MAX_NUMERIC_SCALE))
```
precision 과 scale 중 하나만 invalid 한 경우는 db_value가 손상된 것으로 판단하지 않는 걸까요?

jongmin-won replied:
코드 리뷰 감사합니다.
말씀하신 부분을 반영하여 수정했습니다.

kangmin5505 commented on src/query/numeric_opfunc.c:
```diff
@@ -2716,15 +2742,15 @@ numeric_get_integral_part (const DB_C_NUMERIC num, const int src_prec, const int
-  char dec_str[DB_MAX_NUMERIC_PRECISION * 4];
-  char new_dec_num[DB_MAX_NUMERIC_PRECISION + 1];
+  char dec_str[NUMERIC_MAX_STRING_SIZE];
+  char new_dec_num[(DB_MAX_NUMERIC_SCALE + 4)];
```
DB_MAX_NUMERIC_SCALE + 4 가 의미하는 바를 명시하는 방법을 사용하는 것이 좋을 같습니다.

jongmin-won replied:
네 DB_MAX_NUMERIC_SCALE(252) + 4 의 의미를 주석으로 추가 했습니다.

kangmin5505 commented on src/query/numeric_opfunc.c:
```diff
@@ -3053,161 +3079,556 @@ numeric_coerce_double_to_num (double adouble, DB_C_NUMERIC num, int *prec, int *
-      if (astring[i] == '.')
+      skip = 1;
+
+      if (astring[parse_pos] >= '1' && astring[parse_pos] <= '9')
```
Step 2, Step 3 처럼 current_char = astring[parse_pos]; 를 사용하는 것이 가독성에 더 좋을 거 같습니다.

jongmin-won replied:
네 step1의 `astring[parse_pos]` 해당 부분을 step2, 3 과 동일하게 수정했습니다.

kangmin5505 commented on src/query/numeric_opfunc.c:
```diff
@@ -3053,161 +3079,556 @@ numeric_coerce_double_to_num (double adouble, DB_C_NUMERIC num, int *prec, int *
-	    {			/* sign found */
-	      if (!sign_found)
+	  parse_pos += skip;	/* Skip spaces */
+	  continue;
+	}
+      else if (current_char == ',')
+	{
+	  return DOMAIN_INCOMPATIBLE;
+	}
+      else
+	{
+	  return DOMAIN_INCOMPATIBLE;
+	}
```
else 로 묶어도 될 거 같습니다.

jongmin-won replied:
기존 로직을 다시 확인해보니, 정수 부분에서는 ',' 사용이 가능했습니다.
```
else if (astring[i] == ',')
  {
    /* Accept ',' character on integer part. */
    if (decimal_part)
      {
        ret = DOMAIN_INCOMPATIBLE;
      }
  }
```
따라서 정수의 경우는 ','를 허용하고, 소수의 경우는 에러를 발생시키도록 수정했습니다.
```sql
select cast('123,123.0' as numeric);
   cast('123,123.0' as numeric)
======================
  123123   

select cast('0.123,123' as numeric);
In the command from line 2,
ERROR: Cannot coerce value of domain "character" to domain "numeric".
```

kangmin5505 commented on src/query/numeric_opfunc.c:
```diff
@@ -3053,161 +3079,556 @@ numeric_coerce_double_to_num (double adouble, DB_C_NUMERIC num, int *prec, int *
+    }
+
+  /* Step 3: Parse fractional part */
+  while (parse_pos < astring_length && has_digit)
```
Step 3 루프에서 has_digit의 값이 변하지 않는다면 while 조건에서 계속 확인할 필요가 없을 거 같습니다.

jongmin-won replied:
has_digit을 while 루프에서 계속 확인할 필요가 없어 보여 제거했습니다.

또한 has_digit의 위치도 다음과 같이 수정했습니다.
```
-- AS-IS
else if (current_char == '.')
	{
	  has_digit = true;	/* Case: "0.0000" */
	  ...

-- TO-BE
/* Step 1 */
if (current_char >= '1' && current_char <= '9')
	{
	  has_digit = true;
	  ...
	  
/* Step 2 */ 
if (current_char >= '0' && current_char <= '9' && !trailing_spaces)
	{
	  has_digit = true;
	  ...

/* Step 3 */ 
if (current_char >= '0' && current_char <= '9' && !trailing_spaces)
	{
	  has_digit = true;
```
수정 된 이유는 아래 두 가지가 있습니다.
1. 의미 명확화(정의 일치)
- has_digit는 "숫자를 하나라도 읽었는가"를 나타내는 플래그입니다. 
  - 이는 '유효 숫자' 여부가 아니라, 단순히 입력에 숫자가 존재하는지를 확인하여 NULL 여부를 판별하기 위한 것입니다.
- 기존에는 '.'에서 has_digit = true를 세팅해 의미가 흐려졌습니다. 
- 이제 숫자(0–9)를 읽은 경우에만 true로 설정합니다.
2. 버그 수정(‘.’ 처리)
- 입력이 `'. '`인 경우 결과값은 NULL로 출력됩니다. ('.'에서 has_digit = true로 설정해도 결과는 동일)
- 그러나 기존 로직에서는 analyze_numeric_string() 단계가 아닌 다른 경로에서 NULL 처리가 이루어졌습니다.
- 이를 수정하여 analyze_numeric_string() 단계에서 일관되게 NULL 판정이 이루어지도록 했습니다.
```sql
select cast('.    ' as numeric);
```

kangmin5505 commented on src/query/numeric_opfunc.c:
```diff
@@ -3053,161 +3079,556 @@ numeric_coerce_double_to_num (double adouble, DB_C_NUMERIC num, int *prec, int *
+      else
+	{
+	  return DOMAIN_INCOMPATIBLE;
+	}
```
else 문으로 묶을 수 있을 거 같습니다.

jongmin-won replied:
네 코드 리뷰 감사합니다.
리뷰 내용대로 else 문으로 묶어 소수 부분의 ',' 문자는 에러를 발생시키도록 수정했습니다.

kangmin5505 commented on src/query/numeric_opfunc.c:
```diff
@@ -3053,161 +3079,556 @@ numeric_coerce_double_to_num (double adouble, DB_C_NUMERIC num, int *prec, int *
+	{
+	  frac_digits[frac_count] = current_char;
+	  /* Track non-zero digits */
+	  if (current_char != '0' || valid_zero)
```
valid_zero 변수의 의미가 이전에 소수부에서 0이 아닌 부분이 나왔다라는 것을 의미한다면, valid_zero 변수를 제거하고 직접 (*frac_first_nz < 0) 를 사용하는 것이 더 명확해 보입니다.

jongmin-won replied:
가독성을 위해 valid_zero 변수를 제거하여 수정했으나, 그에 따라 조건이 약간 달라집니다.

*frac_first_nz의 초기값은 -1이므로, 단순히 (*frac_first_nz < 0) 조건만 사용할 경우 아래 예시처럼 소수점 뒤의 0까지 유효한 값으로 인식하게 됩니다.

```sql
select 0.012300;
```

따라서, 최종적으로 아래와 같이 수정했습니다.
```
if (current_char != '0' || *frac_first_nz >= 0)
  {
	if (*frac_first_nz < 0)
...
```

kangmin5505 commented on src/query/numeric_opfunc.c:
```diff
@@ -3053,161 +3079,556 @@ numeric_coerce_double_to_num (double adouble, DB_C_NUMERIC num, int *prec, int *
+  const char *temp_int_digits = NULL, *temp_frac_digits = NULL;
+  int temp_int_len = 0, temp_frac_len = 0;
+  char next_digit = '0';
+  int tmp_prec = 0, tmp_scale = 0;
```
사소한 부분인데, temp와 tmp를 혼용하신 이유가 있을까요?

jongmin-won replied:
특별한 이유는 없어, tmp로 통일했습니다.

kangmin5505 commented on src/query/numeric_opfunc.c:
```diff
@@ -3053,161 +3079,556 @@ numeric_coerce_double_to_num (double adouble, DB_C_NUMERIC num, int *prec, int *
+		  pad_character_zero = false;
+		  valid_zero = true;
 		}
+	      *frac_last_nz = frac_count;
```
frac_last_nz 변수는 (*frac_first_nz >= 0) 일 때, current_char == '0' 이여도 값이 변경되어야 한다면, 변수명이 혼동을 주는 거 같습니다.

jongmin-won replied:
해당 변수명은 "소수 부분의 0이 아닌 숫자"만을 표시한다고 오해될 수 있습니다.

원래는 trailing zero 제거를 위해 사용되던 변수명으로, 예를 들어 0.012300 값에서 123만 확인하려는 목적이었습니다.
그러나 Oracle의 precision 개념과, trailing zero를 제거하지 않기로 한 논의 결과에 따르면 0.012300의 유효 숫자는 12300으로 인식해야 합니다.

따라서 의미를 명확히 하기 위해, "소수 부분의 유효 숫자"를 나타내는 변수명으로 `frac_first_sig_digit`, `frac_last_sig_digit`으로 수정했습니다.

kangmin5505 commented on src/query/numeric_opfunc.c:
```diff
@@ -3053,161 +3079,556 @@ numeric_coerce_double_to_num (double adouble, DB_C_NUMERIC num, int *prec, int *
-      goto exit_on_error;
+      if (temp_int_len == 0)
+	{
+	  scale = frac_zero_cnt == 0 ? 0 : frac_zero_cnt;
```
scale = frac_zero_cnt; 로 충분할 거 같습니다.

jongmin-won replied:
과거 trailing zero제거하기 위해 추가되었던 코드였으나,  
논의 결과 trailing zero를 제거하지 않기로 하여 원복하는 과정에서 일부 누락된 부분이 있었습니다.  
이에 다시 수정하여 아래와 같이 결과가 출력되도록 했습니다.

```sql
SELECT 0.00000000000000000000000000999999999999999999999999999999999999999999999;
-- 0.0000000000000000000000000010000000000000000000000000000000000000

SELECT 0.999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999;
-- 1.0000000000000000000000000000000000000
```

Commit:
1. Applied code review feedback
2. In the domain pack/unpack process, utilized extended_scale (which was not previously used) to handle cases where scale >= 128 and negative scale values
3. In the numeric parsing process, some legacy trailing-zero removal logic remained.
   Since we decided not to remove trailing zeros, this part was revised,
   and additional fixes were made to clean up the remaining inconsistencies

Commit:
numeric_coerce_string_to_num: rename frac_first_nz/frac_last_nz -> frac_first_sig_digit/frac_last_sig_digit for clarity

Commit:
Modified the scale packing/unpacking of domain: previously calculated on a 1-byte basis, now adjusted to fit an int.

Commit:
Updated the comment.

Commit:
Merge remote-tracking branch 'upstream/develop' into CBRD-26243

hyunikn commented:
> 예: -123.45는 숫자(5)+소수점(1)=6자
> 예: +0.0012는 0(1) + 소수점(1) + 숫자(4)= 6자

부호도 포함한다고 했으니, 위 각각 6자가 아니라 7자 인가요?

hyunikn commented:
> 부호가 제외된 이유

위 '리터럴 길이 제한' 절에서는 부호가 포함된다고 되어 있는데, 여기서 제외 된다고 하는 것은 어떤 경우에 제외된다는 뜻인가요.

hyunikn commented:
> (1,127) <= (prec,scale) <= (38,-84)

1 <= prec <= 38, -84 <= scale <= 127 이렇게 쓰는 것이 의도한 바 같습니다. 
위와 같이 쓰려면 순서쌍의 대소 비교가 어떤 의미인지 정의하고 써야 합니다.

jongmin-won commented:
> > 예: -123.45는 숫자(5)+소수점(1)=6자
> > 예: +0.0012는 0(1) + 소수점(1) + 숫자(4)= 6자
> 
> 부호도 포함한다고 했으니, 위 각각 6자가 아니라 7자 인가요?


Lex와 Bison에서 부호를 포함한 리터럴이 그대로 전달될 것으로 예상했으나, 실제로 확인해 보니 String이 아닌 `select -123.45` 형태의 리터럴은 csql_grammar.y의 `unsigned_real` 키워드에서 처리되며, 이때 `val->info.value.text = $1;` 값은 "123.45"로 설정됩니다.

따라서 예제 동작은 정상이며, Purpose의 리터럴 길이 제한 설명 부분에서 부호(+/-)는 길이에 포함되지 않는다고 수정하겠습니다.

```
AS-IS
부호(+/−), "0."(소수점 앞의 0과 소수점)은 포함하여 전체 길이가 254자 이내 까지 허용

TO-BE
부호(+/−)는 길이에 포함되지 않고, "0."(소수점 앞의 0과 소수점)은 길이에 포함되며, 전체 길이는 254자 이내까지만 허용.
```

jongmin-won commented:
> > 부호가 제외된 이유
> 
> 위 '리터럴 길이 제한' 절에서는 부호가 포함된다고 되어 있는데, 여기서 제외 된다고 하는 것은 어떤 경우에 제외된다는 뜻인가요.

아래 예제처럼 query를 수행한 후 numeric_coerce_string_to_num()에 전달되는 astring 값을 확인해 보면, 음수 기호가 제외된 "123.45"만 넘어옵니다.
```sql
select -123.45;
```
이는 음수 표시가 PT_EXPR로 변환되기 때문입니다.
(참고: + 부호는 PT_EXPR이 아닌 PT_VALUE로 바로 변환됩니다.)

반면, 아래와 같이 문자열을 CAST하는 경우에는 astring 값이 "-123.45"로 그대로 전달됩니다.
(+ 부호도 동일하게 "+123.45"로 전달 됩니다.)
```sql
select cast('-123.45' as numeric);
```

jongmin-won commented:
> > (1,127) <= (prec,scale) <= (38,-84)
> 
> 1 <= prec <= 38, -84 <= scale <= 127 이렇게 쓰는 것이 의도한 바 같습니다. 위와 같이 쓰려면 순서쌍의 대소 비교가 어떤 의미인지 정의하고 써야 합니다.

확인 감사합니다.
Prec/Scale 범위는 아래와 같이 정리하여 수정하겠습니다.
```
1 <= prec <= 38
-84 <= scale <= 127
```

hyunikn commented on src/object/object_representation.c:
```diff
@@ -2685,7 +2695,7 @@ or_packed_domain_size (TP_DOMAIN * domain, int include_classoids)
 	}
 
-      if (scale >= OR_DOMAIN_SCALE_MAX)
+      if (scale >= OR_DOMAIN_SCALE_EXT_FLAG || scale < 0)
```
이 경우가 아니면 scale 이 pack 되지 않는 것인가요.
scale 크기가 size 에 더해지는 다른 부분이 안보여서 여쭤봅니다.

jongmin-won replied:
해당 조건이 아니더라도 scale은 항상 pack 됩니다.
- 조건을 만족하는 경우에는 extended_scale 에 값을 별도로 추가하여 전달하며,
- 조건을 만족하지 않는 경우에는 기존과 동일하게 carrier에 직접 scale 값을 담습니다.
```
-- or_put_domain() 의 2808 ~ 2816 라인
if (scale >= OR_DOMAIN_SCALE_EXT_FLAG || scale < 0)
	    {
	      carrier |= OR_DOMAIN_SCALE_MAX << OR_DOMAIN_SCALE_SHIFT;
	      extended_scale = d->scale;
	    }
	  else
	    {
	      carrier |= scale << OR_DOMAIN_SCALE_SHIFT;
	    }

-- or_put_domain() 의 2977 ~ 2984 라인
      if (extended_scale)
	{
	  rc = or_put_int (buf, extended_scale);
	  if (rc != NO_ERROR)
	    {
	      return rc;
	    }
	}
```

hyunikn replied:
carrier 에 담기는 내용의 size 는 포함되지 않는군요.
